include/gridtools/reduction/gpu.hpp

```diff
@@ -0,0 +1,324 @@
+
+#include <cassert>
+#include <cstdlib>
+#include <type_traits>
+
+#include "../common/ct_dispatch.hpp"
+#include "../common/cuda_runtime.hpp"
+#include "../common/cuda_util.hpp"
+#include "../common/host_device.hpp"
+#include "../common/integral_constant.hpp"
+#include "../common/numeric.hpp"
+#include "../meta.hpp"
+#include "functions.hpp"
+
+namespace gridtools {
+    namespace reduction {
+        namespace gpu_backend {
+            struct tiny_kernel {
+                template <class F, class T>
+                GT_FUNCTION_DEVICE void operator()(F f, T const *__restrict__ in, T *__restrict__ out) const {
+                    in += blockIdx.x * 2;
+                    *(out + blockIdx.x) = f(*in, *(in + 1));
+                }
+            };
+
+            template <size_t BlockSize,
+                size_t Step,
+                class F,
+                class T,
+                std::enable_if_t<(BlockSize >= 2 * Step), int> = 0>
+            GT_FUNCTION_DEVICE void shmem_reduce(F f, T *val) {
+                __syncthreads();
+                if (threadIdx.x < Step)
+                    *val = f(*val, *(val + Step));
+            }
+            template <size_t BlockSize,
+                size_t Step,
+                class F,
+                class T,
+                std::enable_if_t<(BlockSize < 2 * Step), int> = 0>
+            GT_FUNCTION_DEVICE void shmem_reduce(F, T *) {}
+
+            template <size_t BlockSize>
+            struct shmem_kernel {
+                static_assert(BlockSize > 1, GT_INTERNAL_ERROR);
+                static_assert(BlockSize <= 1024, GT_INTERNAL_ERROR);
+                template <class F, class T>
+                GT_FUNCTION_DEVICE void operator()(F f, T const *__restrict__ in, T *__restrict__ out) const {
+                    in += blockIdx.x * BlockSize * 2 + threadIdx.x;
+                    __shared__ T buff[BlockSize];
+                    T *val = buff + threadIdx.x;
+                    *val = f(*in, *(in + BlockSize));
+                    shmem_reduce<BlockSize, 512>(f, val);
+                    shmem_reduce<BlockSize, 256>(f, val);
+                    shmem_reduce<BlockSize, 128>(f, val);
+                    shmem_reduce<BlockSize, 64>(f, val);
+                    shmem_reduce<BlockSize, 32>(f, val);
+                    shmem_reduce<BlockSize, 16>(f, val);
+                    shmem_reduce<BlockSize, 8>(f, val);
+                    shmem_reduce<BlockSize, 4>(f, val);
+                    shmem_reduce<BlockSize, 2>(f, val);
+                    __syncthreads();
+                    if (threadIdx.x == 0)
+                        *(out + blockIdx.x) = f(*val, *(val + 1));
+                }
+            };
+
+            // TODO(fthaler): to figure out if __shfl_down_sync is supported on HIP
+            template <class F, class T>
+            GT_FUNCTION_DEVICE T warp_reduce(F f, unsigned mask, T val) {
+#ifdef __HIP__
+                val = f(val, __shfl_down(val, 16));
```

Comment:
Note: there is no mask argument for HIPs `__shfl_down` (https://rocmdocs.amd.com/en/latest/Programming_Guides/Kernel_language.html?highlight=shfl_down#id64), so there is a different strategy required. rocBLAS actually uses a much simpler implementation, but I doubt that this is the ideal one: https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/library/src/blas1/reduction.hpp. I would also be fine with just having a verifying HIP implementation and I could run some benchmarks to see which is actually a good strategy as I haven’t found much documentation there.